GVFS/GVFS.UnitTests/Common/NuGetUpgrade/OrgNuGetUpgraderTests.cs

```diff
@@ -0,0 +1,175 @@
+using GVFS.Common;
+using GVFS.Common.Git;
+using GVFS.Common.NuGetUpgrade;
+using GVFS.Common.Tracing;
+using GVFS.Tests.Should;
+using GVFS.UnitTests.Category;
+using GVFS.UnitTests.Mock.Common;
+using GVFS.UnitTests.Mock.FileSystem;
+using Moq;
+using Moq.Protected;
+using NuGet.Packaging.Core;
+using NuGet.Protocol.Core.Types;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace GVFS.UnitTests.Common.NuGetUpgrade
+{
+    [TestFixture]
+    public class OrgNuGetUpgraderTests
+    {
+        private const string CurrentVersion = "1.5.1185.0";
+        private const string NewerVersion = "1.6.1185.0";
+
+        private const string DefaultUpgradeFeedPackageName = "package";
+        private const string DefaultUpgradeFeedUrl = "https://pkgs.dev.azure.com/contoso/";
+        private const string DefaultOrgInfoServerUrl = "https://www.contoso.com";
+        private const string DefaultRing = "slow";
+
+        private OrgNuGetUpgrader upgrader;
+
+        private MockTracer tracer;
+
+        private OrgNuGetUpgrader.OrgNuGetUpgraderConfig upgraderConfig;
+
+        private Mock<NuGetFeed> mockNuGetFeed;
+        private MockFileSystem mockFileSystem;
+        private Mock<ICredentialStore> mockCredentialManager;
+        private Mock<HttpMessageHandler> httpMessageHandlerMock;
+
+        private string downloadDirectoryPath = Path.Combine(
+            $"mock:{Path.DirectorySeparatorChar}",
+            ProductUpgraderInfo.UpgradeDirectoryName,
+            ProductUpgraderInfo.DownloadDirectory);
+
+        private interface IHttpMessageHandlerProtectedMembers
+        {
+            Task<HttpResponseMessage> SendAsync(HttpRequestMessage message, CancellationToken token);
+        }
+
+        public static IEnumerable<Exception> NetworkFailureCases()
+        {
+            yield return new HttpRequestException("Response status code does not indicate success: 401: (Unauthorized)");
+            yield return new TaskCanceledException("Task canceled");

```

Comment:
Is it also possible to get a `WebException`? (I noticed we're catching that exception in `HttpRequestor.SendRequest`, but I'm not sure it could get thrown in this code path)

Reply:
Good question - according to the msdn documentation, this is not one of the exceptions thrown by HttpClient. It looks like `WebClient.Download` could throw this exception (e.g. as used in `GitHubUpgrader`, but not `HttpClient.GetStringAsync`...